### PR TITLE
Travis scripts for python3 and pytest for cmake.

### PR DESCRIPTION
**Pull request by philkr**
This PR adds python tests for CMake, and allows python3 to be tested on Travis. Later took a bit of hacking to get working:
 * The python tests are updated to be python3 compatible
 * We now have two miniconda directories (```$HOME/miniconda``` for python 2.7 and ```$HOME/miniconda3``` for python3).
 * Boost is installed through miniconda to make sure we have a boost_python that is compatible with the python version used.
 * bc was replaced by awk in the Makefile, as miniconda installs a new libreadline.so that breaks bc.
 * Remove the CPU_ONLY flag for cmake and WITH_CUDA (fixes issue #2722)
 * cmake 3 is required to ensure that the python interpreter and library versions match (the binary is downloaded directory from cmake.org)
 * For python3 protobuf 3 is compiled directly from source

### CODE REVIEW

File: scripts/travis/travis_build_and_test.sh [hunk 1]
```diff
@@ -7,8 +7,17 @@ MAKE="make --jobs=$NUM_THREADS --keep-going"
 if $WITH_CMAKE; then
   mkdir build
   cd build
-  cmake -DBUILD_python=ON -DCMAKE_BUILD_TYPE=Release -DCPU_ONLY=ON ..
```
Comment (shelhamer) [inline scripts/travis/travis_build_and_test.sh]: Can you split this fix into another commit or amend the message to note the fix?

File: python/caffe/test/test_net.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 import tempfile
 import os
 import numpy as np
+import six
```
Comment (shelhamer) [inline python/caffe/test/test_net.py]: Please add six to the requirements.txt since it isn't always bundled as far as I know.

Comment (philkr) [inline python/caffe/test/test_net.py]: already done in #2714

File: scripts/travis/travis_build_and_test.sh (hunk 1 above)
Comment (philkr) [inline scripts/travis/travis_build_and_test.sh]: amended the message.

File: python/caffe/test/test_net.py (hunk 2 above)
Comment (shelhamer) [inline python/caffe/test/test_net.py]: Oh right, thanks for the reminder.
